### PR TITLE
chore: add new methods in VisbilityBypasser to simplify "this or that" usage pattern.

### LINKED ISSUE #1797: Refactor VisibilityBypasser
We have a few places in the code where we either want to find property/method A *or* property/method B, depending on which one is available. Right now the only way to do that is with a try/catch on the VisibilityBypasser call. It would be cleaner if we could either have methods to test if it exists first, or have alternate versions that return null/a status code if the field isn't available. Or maybe even a function that says "give me A or B, whichever exists".

### Acceptance Criteria
New functionality and integration or unit tests.

### Estimates
S/M

### PR DESCRIPTION
**Pull request by tippmar-nr**
Adds a couple of `TryXxx()` methods in `VisibilityBypasser` that simplify calling patterns where we attempt to get a particular property or method, then fall back to getting a different property or method. 

Resolves #1797.

### CODE REVIEW

File: src/Agent/NewRelic/Agent/Extensions/Providers/Wrapper/Log4NetLogging/Log4netWrapper.cs [hunk 1]
```diff
@@ -59,32 +59,18 @@ private void RecordLogMessage(object logEvent, Type logEventType, IAgent agent)
...
-
-            try
-            {
-                getLogExceptionFunc = _getLogException ??= VisibilityBypasser.Instance.GeneratePropertyAccessor<Exception>(logEventType, "ExceptionObject");
-            }
-            catch
+            if (_getLogException == null
+                && !VisibilityBypasser.Instance.TryGeneratePropertyAccessor<Exception>(logEventType, "ExceptionObject", out _getLogException)
+                // Legacy property, mainly used by Sitecore
+                && !VisibilityBypasser.Instance.TryGeneratePropertyAccessor<Exception>(logEventType, "m_thrownException", out _getLogException))
```
Comment (nrcventura) [inline src/Agent/NewRelic/Agent/Extensions/Providers/Wrapper/Log4NetLogging/Log4netWrapper.cs:65]: Even though this is more compact and it reduces nesting, I find this harder to read. It is easier for me to understand if there is only 1 method with an out parameter being called in the conditional.

**Comment by chynesNR** [comment]
What I was imagining when I wrote the ticket was to avoid the overhead of a caught exception at all, since we're the ones throwing it, if we can't find the given method. Here we're still relying on `GetMethodInfo` to throw when it can't find the first one in our "this or that" scenario. That's the refactoring I was talking about, but if it ends up being a lot of rework it's probably not worth it.

**Comment by tippmar-nr** [comment]
@chynesNR Yeah, I see what you're saying - but it gets pretty complicated to unravel the "throw an exception if we can't find the thing" logic, as that's how every method in VisibilityBypasser was originally implemented. So my refactor at least helps isolate the ugliness from the outside world, I guess... ;-)

File: src/Agent/NewRelic/Agent/Extensions/Providers/Wrapper/Log4NetLogging/Log4netWrapper.cs (hunk 1 above)
Comment (tippmar-nr) [reply src/Agent/NewRelic/Agent/Extensions/Providers/Wrapper/Log4NetLogging/Log4netWrapper.cs:65]: Would you prefer something like this? 
```
            if (_getLogException == null)
            {
                if (!VisibilityBypasser.Instance.TryGeneratePropertyAccessor<Exception>(logEventType, "ExceptionObject", out _getLogException))
                {
                    // Legacy property, mainly used by Sitecore
                    if (!VisibilityBypasser.Instance.TryGeneratePropertyAccessor<Exception>(logEventType, "m_thrownException", out _getLogException))
                    {
                        _getLogException = (x) => null;
                    }
                }
            }
```

Comment (nrcventura) [reply src/Agent/NewRelic/Agent/Extensions/Providers/Wrapper/Log4NetLogging/Log4netWrapper.cs:65]: I think that makes it clearer.

**Comment by nrcventura** [comment]
> @chynesNR Yeah, I see what you're saying - but it gets pretty complicated to unravel the "throw an exception if we can't find the thing" logic, as that's how every method in VisibilityBypasser was originally implemented. So my refactor at least helps isolate the ugliness from the outside world, I guess... ;-)

I think that the refactor could be done in multiple steps. This first PR cleans up the usage. The follow up PR could update the code to not rely on the exceptions.

**Review by jaffinito — APPROVED** [review]
I don't care for using exceptions as logic, but I am not seeing a way not to here.  Looks good!

**Pull request merged**